tests/std/tests/P3349R1_contiguous_iterators_to_pointers/test.cpp

```diff
@@ -0,0 +1,463 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <format>
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <numeric>
+#include <ranges>
+#include <string>
+#include <vector>
+
+using namespace std;
+
+struct safe_iter_out_of_bounds_err {};
+
+enum range_type { range_large = 3, range_medium = 4, range_small = 5 };
+
+bool safe_iter_nothrow_OOB_sign = false;
+
+// When `nothrow` is true, disables bounds checking during iterator increment/decrement (to verify `to_address` calls),
+// logging out-of-bounds accesses instead of throwing. When false, throws immediately on bounds violation.
+template <class T, bool nothrow = false>
+class safe_iter {
+public:
+    using iterator_concept  = contiguous_iterator_tag;
+    using iterator_category = random_access_iterator_tag;
+    using value_type        = T;
+    using difference_type   = ptrdiff_t;
+    using pointer           = T*;
+    using reference         = T&;
+
+    constexpr safe_iter(pointer ptr, pointer range_first, pointer range_last) noexcept
+        : current_ptr{ptr}, range_first_ptr{range_first}, range_last_ptr{range_last} {
+        assert(current_ptr >= range_first_ptr && current_ptr <= range_last_ptr);

```

Comment:
It's better to assert one condition at a time, so that if the assertion fails, we'll see in the test log what the problem was.